composition-js/src/merging/merge.ts

```diff
@@ -3606,4 +3689,344 @@ class Merger {
+          } catch (e) {
+            if (!(e instanceof GraphQLError)) {
+              throw e;
+            }
+            // target subgraph info should always be provided but just in case
+            const enumSubgraphValue = joinDirectiveOnFromContext.arguments().graph;
+            const subgraph = enumSubgraphValue ? this.joinSpecNamesToSubgraphNames.get(enumSubgraphValue) : undefined;
+            if (subgraph) {
+              errors.push(addSubgraphToError(e, subgraph));
+            } else {
+              errors.push(e);
+            }
+          }
+        }
+      }
+    }
+    return errors;
+  }
+
+  private authRequirementsOnElement(element: NamedType | FieldDefinition<ObjectType>): AuthRequirementsOnElement | undefined {
+    const requirements = new AuthRequirementsOnElement();
+    if (this.authenticatedDirective) {
+      const appliedDirective = element.appliedDirectivesOf(this.authenticatedDirective)?.[0];
+      if (appliedDirective) {
+        requirements.isAuthenticated = true;
+      }
+    }
+
+    if (this.requiresScopesDirective) {
+      const appliedDirective = element.appliedDirectivesOf(this.requiresScopesDirective)?.[0];
+      if (appliedDirective) {
+        const { scopes } = appliedDirective.arguments();
+        requirements.scopes = scopes;
+      }
+    }
+
+    if (this.policyDirective) {
+      const appliedDirective = element.appliedDirectivesOf(this.policyDirective)?.[0];
+      if (appliedDirective) {
+        const { policies } = appliedDirective.arguments();
+        requirements.policies = policies;
+      }
+    }
+
+    if (requirements.isAuthenticated || requirements.scopes || requirements.policies) {
+      return requirements;
+    } else {
+      return;
+    }
+  }
+
+  private verifyAuthRequirementsOnSelectionSet(authRequirements: AuthRequirements, selectionSet: SelectionSet) {
+    const parentType = this.schema.type(selectionSet.parentType.name);
+    for (const selection of selectionSet.selections()) {
+      if (selection instanceof FieldSelection) {
+        if (parentType instanceof InterfaceType) {
+          // if we are referencing an interface field we need to check against all implementations
+          for (const implType of parentType.possibleRuntimeTypes()) {
+            const requirementsOnImplType = this.authRequirementsOnElement(implType);
+            if (!authRequirements.satisfies(requirementsOnImplType)) {
+              const msg = `Field "${authRequirements.fieldCoordinate}" does not specify necessary @authenticated, @requiresScopes `
+                  + `and/or @policy auth requirements to access the transitive interface "${parentType}" data from ${authRequirements.directive} selection set.`;
+              throw ERRORS.MISSING_TRANSITIVE_AUTH_REQUIREMENTS.err(msg);
+            }
+            // re-check the implementation auth condition
+            this.verifyAuthOnFieldSelection(implType, selection, authRequirements);
+          }
```

Comment:
Since we've upwards-propagated auth requirements by this point, I don't think there's a reason to type explode here; it should be fine to just check the auth requirements on the interface field and the field's return type, before recursing into the field's selection set (if any).

It's particularly important not to type-explode in this case, since type explosion may lead to differing types downstream in the query due to output type covariance, and those types may have less-strict requirements than the abstract types they implement.

Reply:
Yeah this was added when were not going to propagate auth reqs. This is redundant now. I'll remove it.